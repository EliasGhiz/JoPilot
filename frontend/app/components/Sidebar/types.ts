import { ReactNode } from 'react';
import type { PaletteMode } from "@mui/material";
import { ThemeVariant } from '../../theme/themeColors';

// Define the navigation item type
export interface NavItem {
  label: string;
  route: string;
  icon: ReactNode;
  group: 'top' | 'bottom';
  // New optional flag: if true then SidebarItem should use a custom logout color.
  useCustomColor?: boolean;
}

export interface SidebarItemProps {
  item: NavItem;
  open: boolean;
  colorMode: PaletteMode;
  themeVariant: ThemeVariant;
}

export interface SidebarProps {
  open: boolean;
  themeVariant: ThemeVariant;
  colorMode: PaletteMode;
}
