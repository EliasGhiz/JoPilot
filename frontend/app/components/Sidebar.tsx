import { Drawer, Box, List, Divider, ListItem, ListItemIcon, ListItemText, Tooltip, Typography } from "@mui/material";
import DashboardIcon from "@mui/icons-material/Dashboard";
import SettingsIcon from "@mui/icons-material/Settings";
import PestControlRodentIcon from "@mui/icons-material/PestControlRodent";
import { useNavigate, useLocation } from "react-router-dom";
import { appVersion } from "../../src/version";
import { getSidebarIconColor } from "../../src/themeConfig";
import { LAYOUT } from "../constants/layout";
import { cloneElement } from "react";
import type { ReactElement } from "react";

// Reusable component for a sidebar navigation item.
const SidebarItem = ({
	icon,
	label,
	route,
	open,
	colorMode,
	baseShade,
	iconSelectedColor
}: {
	icon: ReactElement;
	label: string;
	route: string;
	open: boolean;
	colorMode: "dark" | "light";
	baseShade: string;
	iconSelectedColor: string;
}) => {
  const navigate = useNavigate();
  const location = useLocation();
  const isSelected = location.pathname === route;
  return (
    <Tooltip title={label} placement="right" disableHoverListener={open}>
      <ListItem
        component="button"
        onClick={() => navigate(route)}
        sx={{
          height: 48,
          alignItems: "center",
          pl: "16px",
          width: "100%",
          cursor: "pointer",
          "&:hover": { bgcolor: isSelected ? baseShade : "action.hover" },
          bgcolor: isSelected ? baseShade : "inherit",
          border: "none",
          outline: "none",
          "&:focus": { outline: "none" }
        }}
      >
        <ListItemIcon sx={{ minWidth: "40px", mr: 0 }}>
          {cloneElement(icon as React.ReactElement<any>, { style: { color: isSelected ? iconSelectedColor : undefined } })}
        </ListItemIcon>
        {open && (
          <ListItemText primary={label} sx={{ ml: 0, color: colorMode === "dark" ? "#FFFFFF" : "inherit" }} />
        )}
      </ListItem>
    </Tooltip>
  );
};

interface SidebarProps {
  open: boolean;
  themeIcon: string;
  colorMode: "dark" | "light";
}

export default function Sidebar({ open, themeIcon, colorMode }: SidebarProps) {
  const collapsedWidth = LAYOUT.collapsedSidebarWidth;
  const drawerWidth = LAYOUT.expandedSidebarWidth;
  const baseShade = colorMode === "dark" ? "#2e2e2e" : "#e0e0e0";
  const iconSelectedColor = getSidebarIconColor(colorMode, themeIcon);

  // Define navigation items
  const navItems = [
    { label: "Dashboard", route: "/dashboard", icon: <DashboardIcon /> },
    { label: "Settings", route: "/settings", icon: <SettingsIcon /> },
    { label: "Test", route: "/test", icon: <PestControlRodentIcon /> }
  ];

  return (
    <Drawer
      variant="permanent"
      anchor="left"
      open={open}
      sx={{
        position: 'fixed',
        left: 0,
        top: 64, // appBarHeight
        width: open ? drawerWidth : collapsedWidth,
        flexShrink: 0,
        "& .MuiDrawer-paper": {
          width: open ? drawerWidth : collapsedWidth,
          transition: "width 0.3s",
          boxSizing: "border-box",
          top: 64, // appBarHeight
          height: `calc(100vh - 64px)`,
          overflowX: "hidden",
          overflowY: "hidden",
          bgcolor: "background.paper",
          color: "text.primary",
        },
      }}
    >
      <Box sx={{ display: "flex", flexDirection: "column", height: "100%" }}>
        <List sx={{ flexGrow: 1 }}>
          {navItems.map(item => (
            <SidebarItem
              key={item.route}
              icon={item.icon}
              label={item.label}
              route={item.route}
              open={open}
              colorMode={colorMode}
              baseShade={baseShade}
              iconSelectedColor={iconSelectedColor}
            />
          ))}
        </List>
        <Divider sx={{ my: 0, bgcolor: "#424242" }} />
        <Box sx={{ textAlign: "center", p: 1 }}>
          <Typography variant="caption" sx={{ color: colorMode === "dark" ? "#FFFFFF" : "inherit", whiteSpace: "nowrap" }}>
            {open ? `Version ${appVersion}` : `v${appVersion}`}
          </Typography>
        </Box>
      </Box>
    </Drawer>
  );
}
