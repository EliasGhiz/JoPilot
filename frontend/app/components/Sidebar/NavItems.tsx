// NavItems.tsx â€“ Exports the list of navigation items for the sidebar.

import DashboardIcon from '@mui/icons-material/Dashboard';
import SettingsIcon from '@mui/icons-material/Settings';
import PestControlRodentIcon from '@mui/icons-material/PestControlRodent';
import LogoutIcon from '@mui/icons-material/Logout';
import FeedIcon from '@mui/icons-material/Feed';
import DescriptionIcon from '@mui/icons-material/Description';
import { NavItem } from './types';
import { AttachFile } from '@mui/icons-material';

// Sidebar Navigation items
export const NAV_ITEMS: NavItem[] = [
  {
    label: 'Dashboard',
    route: '/dashboard',
    icon: <DashboardIcon />,
    group: 'top'
  },
  {
    label: 'Applications',
    route: '/applications',
    icon: <DescriptionIcon />,
    group: 'top'
  },
  {
    label: 'Job Postings',
    route: '/postings',
    icon: <FeedIcon />,
    group: 'top'
  },
  {
    label: 'Analyze Resume',
    route: '/analyze',
    icon: <AttachFile />, // Replace with a relevant icon
    group: 'top'
  },
  {
    label: 'Settings',
    route: '/settings',
    icon: <SettingsIcon />,
    group: 'bottom'
  },
  {
    label: 'Test',
    route: '/test',
    icon: <PestControlRodentIcon />,
    group: 'bottom'
  },
  {
    label: 'Logout',
    route: '/logout',
    icon: <LogoutIcon />,
    group: 'bottom',
    // New: use custom logout color from themeColors
    useCustomColor: true
  }
];
