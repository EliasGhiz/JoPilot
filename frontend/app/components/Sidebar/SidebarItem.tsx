// SidebarItem.tsx â€“ Renders individual navigation items in the sidebar with selection state and icons.

import { cloneElement } from 'react';
import {
  ListItemButton,
    ListItemText,
  Tooltip,
  Box
} from '@mui/material';
import { useNavigate, useLocation } from 'react-router-dom';
import { getSidebarIconColor, getThemeColor } from 'app/theme/colorSystem';
import { SidebarItemProps } from './types';

// Sidebar item component
export const SidebarItem = ({ item, open, colorMode, themeVariant }: SidebarItemProps) => {
  const navigate = useNavigate();
  const location = useLocation();
  const isSelected = location.pathname === item.route;
  
  // Subtle background shade hint for selection
  const baseShade = colorMode === 'dark' 
    ? 'rgba(255, 255, 255, 0.15)' 
    : 'rgba(0, 0, 0, 0.09)';
  
  const iconColor = isSelected
    ? (themeVariant === 'blue'
          ? getThemeColor('blue', 'primary', colorMode, 45)
          : themeVariant === 'red'
            ? getThemeColor('red', 'primary', colorMode, 50)
            : themeVariant === 'gray'
              ? getThemeColor('gray', 'neutral', colorMode, colorMode === 'light' ? 30 : 90)
              : getSidebarIconColor(colorMode, themeVariant))
    : (colorMode === 'dark'
          ? getThemeColor('gray', 'neutral', colorMode, 70)
          : getThemeColor('gray', 'neutral', colorMode, 40));
  
  return (
    <Tooltip title={item.label} placement="right" disableHoverListener={open}>
      <ListItemButton
        onClick={() => navigate(item.route)}
        sx={{
          // Different styling for open vs closed sidebar
          p: 0,
          my: 0.5,
          // Don't touch  horizontal margins when collapsed so the icon container stays fixed
          mx: 0,
          height: 48,
          borderRadius: open ? '4px' : 0,
          bgcolor: isSelected ? baseShade : 'inherit',
          transition: 'none',
          '&:hover': { 
            bgcolor: isSelected 
              ? baseShade 
              : (theme) => theme.palette.mode === 'dark' 
                ? 'rgba(255, 255, 255, 0.08)' 
                : 'rgba(0, 0, 0, 0.04)' 
          },
          display: 'flex',
          alignItems: 'center',
          overflow: 'hidden',
        }}
      >
        {/* Fixed container for the icon */}
        <Box 
          sx={{
            width: 56, // Fixed width (always 56px)
            height: 48,
            display: 'flex',
            alignItems: 'center',
            justifyContent: 'center',
            flexShrink: 0,
          }}
        >
          {cloneElement(item.icon as React.ReactElement<any>, { 
            style: { 
              color: iconColor,
              fontSize: '1.4rem',
            }
          })}
        </Box>
        
        {/* The text - only shown when sidebar is open */}
        {open && (
          <ListItemText 
            primary={item.label} 
            sx={{
              ml: 1,
              color: colorMode === 'dark' 
                ? getThemeColor('gray', 'neutral', colorMode, 95) 
                : 'inherit'
            }}
          />
        )}
      </ListItemButton>
    </Tooltip>
  );
};
