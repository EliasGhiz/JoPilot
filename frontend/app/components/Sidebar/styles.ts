// styles.ts â€“ Sidebar styles centralized definitions.

import type { PaletteMode } from "@mui/material";
import { getThemeColor } from "app/theme/colorSystem";
import { TRANSITION_SPEED_FAST } from "app/theme/uiConstants";

// Styling objects to reduce inline styles
export const styles = {
  itemText: (colorMode: PaletteMode) => ({
    ml: 0, 
    color: colorMode === 'dark' ? getThemeColor('gray', 'neutral', colorMode, 95) : 'inherit'
  }),
  
  versionText: (colorMode: PaletteMode) => ({
    color: colorMode === 'dark' 
      ? getThemeColor('gray', 'neutral', colorMode, 60)
      : getThemeColor('gray', 'neutral', colorMode, 40),
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    display: 'block'
  }),
  
  drawer: (open: boolean, drawerWidth: number, collapsedWidth: number, appBarHeight: number) => ({
    position: 'fixed',
    width: open ? drawerWidth : collapsedWidth,
    flexShrink: 0,
    '& .MuiDrawer-paper': {
      width: open ? drawerWidth : collapsedWidth,
      transition: `width ${TRANSITION_SPEED_FAST}`,
      top: appBarHeight,
      height: `calc(100vh - ${appBarHeight}px)`,
      overflowX: 'hidden',
      overflowY: 'hidden',
      bgcolor: 'background.paper',
      p: 0
    },
    '& .MuiList-root': {
      p: 0,
      pt: 1
    }
  })
};
