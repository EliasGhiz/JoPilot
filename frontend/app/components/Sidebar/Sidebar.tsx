// Sidebar.tsx â€“ Renders the application's side navigation drawer.

import {
  Drawer,
  List,
  Typography,
  Box,
  Divider,
  useTheme
} from '@mui/material';
import { appVersion } from 'app/utils/version';
import { styles } from './styles';
import { SidebarProps } from './types';
import { SidebarItem } from './SidebarItem';
import { NAV_ITEMS } from './NavItems';
import { getThemeColor } from '../../theme/colorSystem';

export default function Sidebar({ open, themeVariant, colorMode }: SidebarProps) {
  const theme = useTheme();
  const { collapsedSidebarWidth: collapsedWidth, expandedSidebarWidth: drawerWidth, appBarHeight } = theme.layout;
  
  // Render navigation items by group
  const renderNavItems = (group: 'top' | 'bottom') => (
    NAV_ITEMS
      .filter(item => item.group === group)
      .map(item => (
        <SidebarItem 
          key={item.route} 
          item={item} 
          open={open} 
          colorMode={colorMode} 
          themeVariant={themeVariant}
        />
      ))
  );

  // Get sidebar divider color using the color system
  const dividerColor = colorMode === 'dark'
    ? getThemeColor(themeVariant, 'neutralVariant', colorMode, 20)
    : getThemeColor(themeVariant, 'neutralVariant', colorMode, 80);

  return (
    <Drawer
      variant="permanent"
      anchor="left"
      sx={styles.drawer(open, drawerWidth, collapsedWidth, appBarHeight)}
    >
      <Box sx={{ 
        display: 'flex', 
        flexDirection: 'column', 
        height: '100%',
        // Remove horizontal padding that may affect icon centering
        px: 0,
      }}>
        {/* Top navigation */}
        <List sx={{ p: 0, pt: 1 }}>{renderNavItems('top')}</List>
        
        {/* Bottom section */}
        <Box sx={{ mt: 'auto' }}>
          <List sx={{ p: 0 }}>{renderNavItems('bottom')}</List>
          <Divider sx={{ backgroundColor: dividerColor }} />
          <Box sx={{ textAlign: 'center', p: 1 }}>
            <Typography variant="caption" sx={styles.versionText(colorMode)}>
              {open ? `Version ${appVersion}` : `v${appVersion}`}
            </Typography>
          </Box>
        </Box>
      </Box>
    </Drawer>
  );
}
