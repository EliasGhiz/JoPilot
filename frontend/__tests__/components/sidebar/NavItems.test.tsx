import React from 'react';
import { render } from '@testing-library/react';
import { NAV_ITEMS } from '../../../app/components/Sidebar/NavItems';

describe('NAV_ITEMS Snapshot Test', () => {
  it('should match the snapshot', () => {
    const { asFragment } = render(
      <div>
        {NAV_ITEMS.map(item => (
          <div key={item.label}>
            {item.icon}
            <span>{item.label}</span>
          </div>
        ))}
      </div>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should contain correct labels and routes', () => {
    expect(NAV_ITEMS).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ label: 'Dashboard', route: '/dashboard' }),
        expect.objectContaining({ label: 'Settings', route: '/settings' }),
        expect.objectContaining({ label: 'Test', route: '/test' }),
      ])
    );
  });
});