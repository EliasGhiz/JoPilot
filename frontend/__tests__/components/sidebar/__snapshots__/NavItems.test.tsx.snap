// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NAV_ITEMS Snapshot Test should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="DashboardIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M3 13h8V3H3zm0 8h8v-6H3zm10 0h8V11h-8zm0-18v6h8V3z"
        />
      </svg>
      <span>
        Dashboard
      </span>
    </div>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="SettingsIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19.14 12.94c.04-.3.06-.61.06-.94 0-.32-.02-.64-.07-.94l2.03-1.58c.18-.14.23-.41.12-.61l-1.92-3.32c-.12-.22-.37-.29-.59-.22l-2.39.96c-.5-.38-1.03-.7-1.62-.94l-.36-2.54c-.04-.24-.24-.41-.48-.41h-3.84c-.24 0-.43.17-.47.41l-.36 2.54c-.59.24-1.13.57-1.62.94l-2.39-.96c-.22-.08-.47 0-.59.22L2.74 8.87c-.12.21-.08.47.12.61l2.03 1.58c-.05.3-.09.63-.09.94s.02.64.07.94l-2.03 1.58c-.18.14-.23.41-.12.61l1.92 3.32c.12.22.37.29.59.22l2.39-.96c.5.38 1.03.7 1.62.94l.36 2.54c.05.24.24.41.48.41h3.84c.24 0 .44-.17.47-.41l.36-2.54c.59-.24 1.13-.56 1.62-.94l2.39.96c.22.08.47 0 .59-.22l1.92-3.32c.12-.22.07-.47-.12-.61zM12 15.6c-1.98 0-3.6-1.62-3.6-3.6s1.62-3.6 3.6-3.6 3.6 1.62 3.6 3.6-1.62 3.6-3.6 3.6"
        />
      </svg>
      <span>
        Settings
      </span>
    </div>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="PestControlRodentIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m21.31 17.38-2.39-2.13c.52-2.36-1.36-4.25-3.42-4.25-1.16 0-3.5.9-3.5 3.5 0 .97.39 1.84 1.03 2.47l-.71.71C11.5 16.87 11 15.74 11 14.5c0-1.7.96-3.17 2.35-3.93-.7-.36-1.48-.57-2.28-.57-2.38 0-4.37 1.65-4.91 3.87C4.91 13.5 4 12.36 4 11c0-1.66 1.34-3 3-3h2.5C10.88 8 12 6.88 12 5.5S10.88 3 9.5 3H8c-.55 0-1 .45-1 1s.45 1 1 1h1.5c.28 0 .5.22.5.5s-.22.5-.5.5H7c-2.76 0-5 2.24-5 5 0 2.42 1.72 4.44 4 4.9v.03C6 18.73 8.27 21 11.07 21h8.86c1.87 0 2.81-2.34 1.38-3.62M18 19c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1"
        />
      </svg>
      <span>
        Test
      </span>
    </div>
    <div>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
        data-testid="LogoutIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="m17 7-1.41 1.41L18.17 11H8v2h10.17l-2.58 2.58L17 17l5-5zM4 5h8V3H4c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h8v-2H4z"
        />
      </svg>
      <span>
        Logout
      </span>
    </div>
  </div>
</DocumentFragment>
`;
